### PR TITLE
add docker-compose documentation for deploying on aws/gcp/digital ocean

### PR DESCRIPTION
**Pull request by ggilmore**
# Overview

This PR adds documentation for deploying docker-compose on each of the major cloud providers: AWS, GCP, and Digital Ocean. Each set of instructions gives recommendations on what machine size to choose, how much storage to give the machine, and documentation pointers about how to create backups of their Sourcegraph instance. 

This PR also tells people what deployment option they should choose when they're starting out: 

| Deployment Type                               | Suggested for                                       | Setup time | Multi-machine? | Auto healing? | Monitoring? |
|-----------------------------------------------|-----------------------------------------------------|------------|----------------|---------------|-------------|
| [Single-container server](../docker/index.md) | Local testing                                       | 60 seconds | Impossible     | No            | No          |
| [Docker-compose](index.md)                    | Small & medium production deployments               | 5 minutes  | Possible       | No            | Yes         |
| [Kubernetes](../cluster.md)                   | Medium & large highly-available cluster deployments | 30 minutes | Easily         | Yes           | Yes         |

A later PR will give instructions for migrating off of `sourcegraph/server` to Docker Compose. 

### Startup scripts

Each of the cloud provider docs provides a startup script to initialize the instance. Each of the scripts follows this general flow:

1. Install `git`
1. Clone https://github.com/sourcegraph/deploy-sourcegraph-docker to the installation directory (AWS: `/home/ec2-user/deploy-sourcegraph-docker`, GCP/DO: `/root/deploy-sourcegraph-docker`) and checkout the latest [release](https://github.com/sourcegraph/deploy-sourcegraph-docker/releases)
1. Prepare the external disk (used for storing all of Docker's data) and mount it into the file system
	- Using the external disk's device name provided in the startup script, check to see if the disk already has an existing  file system
	   - If it's **unformatted**, then format with `xfs` for AWS, and `ext4` for GCP/DO
       - If the disk **already has a filesystem**, do nothing (this prevents data from being destroyed if the user is using a disk that had a backup restored onto it).  
	   - Notes:
			- The same init script can be used on both the initial install and when spinning up a VM that has existing sourcegraph data on it.
            - Digital Ocean's webUI doesn't allow you to attach an existing volume to a new droplet during the creation process - you can only do so **after** the instance is initialized. I'll need to investigate instructions for how to spin up a backup on DO (simply run the init script after you login for the first time)?
	- Add the external disk's UUID to `/etc/fstab` so that the disk is automatically mounted upon reboots.
       - A lot of cloud provider documentation said that you can't rely on the device name being stable between reboots. Some cloud providers allow you to give names to disks, but the resulting label in `/dev/` wasn't predictable in my experience. I think using the device name is fine for now since it's only used for the initial format and the UUID is what actually gets stored in `/etc/fstab`.
1. Install docker
1. Install jq
1. Use jq modify `/etc/docker/daemon.json` to change the [`data-root`](https://docs.docker.com/engine/reference/commandline/dockerd/) setting to point to `/mnt/docker-data`. 
	- All of docker-compose's data is stored via docker volumes. Telling Docker to store all of its data under `/mnt/docker-data` allows customer's to easily snapshot/backup all of the user data in a similar workflow to what they have with `sourcegraph/server`. 
1. Restore the docker daemon to pickup these changes
1. Install docker-compose 
1. Run `docker-compose up -d` to start the sourcegraph/services and restart them on reboots 

### Backups

Docker Compose doesn't offer any built-in tool to mass backup Docker volumes, and 
[Docker's documentation](https://docs.docker.com/storage/volumes/#backup-restore-or-migrate-data-volumes) only offers instructions for how to create/restore a tarball of a container on the individual container level. (I haven't tested this yet, but going down this route would involve custom scripting and would require managing these tarballs somewhere such as an S3 bucket, etc.)

As mentioned above, most of our customers currently solve this by mounting an external disk to the ~/.sourcegraph directory.  I have replicated this workflow by telling Docker to store all of its data (including volumes) under `/mnt/docker-data`. The docs also contain pointers for how to take snapshots of the external disk. 

In addition to the above method, I also call out that using an external Postgres service is an easy way to backup all of Sourcegraph's user data. 

---

Notes:

I have personally tested these docs for AWS, GCP, and DO. I have also tested restoring backups on AWS and GCP (and not DO for reasons stated above). 

I would still like someone to take an in depth look at this PR, the wording, and actually test out my instructions, but this PR needs to be merged already so that customers and take a look.

### CODE REVIEW

File: doc/admin/install/docker-compose/index.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# Install Sourcegraph with Docker Compose
+
+When running Sourcegraph in production, deploying Sourcegraph via [Docker Compose](https://docs.docker.com/compose/) is the default installation method that we recommend. However:
```
Comment (slimsag) [inline doc/admin/install/docker-compose/index.md]:
```suggestion
| Deployment Type              | Suggested for                                       | Setup time | Multi-machine? | Auto healing? | Monitoring? |                                                                                                                                 |
|------------------------------|-----------------------------------------------------|------------|----------------|---------------|-------------|
| [Single-container server](#) | Local testing                                       | 60 seconds | Impossible     | No            | No          |
| [Docker-compose](#)          | Small & medium production deployments               | 5 minutes  | Possible       | No            | Yes         |
| [Kubernetes](#)              | Medium & large highly-available cluster deployments | 30 minutes | Easily         | Yes           | Yes         |
```

Comment (slimsag) [reply doc/admin/install/docker-compose/index.md]: I propose the table above to explain to users how they should choose (also, maybe it shouldn't be explained in this file but rather at a higher level? i.e. the file that links to this one)

Here is how it looks rendered:

| Deployment Type              | Suggested for                                       | Setup time | Multi-machine? | Auto healing? | Monitoring? |
|------------------------------|-----------------------------------------------------|------------|----------------|---------------|-------------|
| [Single-container server](#) | Local testing                                       | 60 seconds | Impossible     | No            | No          |
| [Docker-compose](#)          | Small & medium production deployments               | 5 minutes  | Possible       | No            | Yes         |
| [Kubernetes](#)              | Medium & large highly-available cluster deployments | 30 minutes | Easily         | Yes           | Yes         |

File: doc/admin/install/docker-compose/index.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+|-----------------------------------------------|-----------------------------------------------------|------------|----------------|---------------|-------------|
+| [Single-container server](../docker/index.md) | Local testing                                       | 60 seconds | Impossible     | No            | No          |
+| [Docker-compose](index.md)                    | Small & medium production deployments               | 5 minutes  | Possible       | No            | Yes         |
+| [Kubernetes](../cluster.md)                   | Medium & large highly-available cluster deployments | 30 minutes | Easily         | Yes           | Yes         |
+
+* If you're just starting out, we recommend [running Sourcegraph locally](../docker/index.md). It takes only a few minutes and lets you try out all of the features.
+* If you need scalability and high-availability beyond what a single-node [Docker Compose](https://docs.docker.com/compose/) can offer, use the [Kubernetes cluster deployment option](https://github.com/sourcegraph/deploy-sourcegraph), instead.
+
+It takes less than 5 minutes to run and install Sourcegraph using Docker Ccompose:
```
Comment (slimsag) [inline doc/admin/install/docker-compose/index.md]:
```suggestion
It takes less than 5 minutes to run and install Sourcegraph using Docker Compose:
```

**Comment by ggilmore** [comment]
@slimsag PTAL

File: doc/admin/install/index.md [hunk 3]
```diff
@@ -1,4 +1,7 @@
 # Installing Sourcegraph
 
-- [Install Sourcegraph with Docker](docker/index.md) **(recommended, easiest)**
```
Comment (dadlerj) [inline doc/admin/install/index.md]: Shouldn't these links still be at the bottom of the page?

File: doc/admin/install/index.md [hunk 4]
```diff
@@ -1,4 +1,7 @@
...
-- [Install Sourcegraph on a cluster](cluster.md)
+| Deployment Type                                       | Suggested for                                       | Setup time | Multi-machine? | Auto healing? | Monitoring? |
```
Comment (dadlerj) [inline doc/admin/install/index.md]: I think this is the right page for this table to be located on. Perhaps take the commentary from the other pages (e.g. "* If you're just starting out...") and put it here. 

Then, remove the charts from all of the other pages, and instead, have a ">NOTE:" (or whatever the syntax is for this) at the top of each of those pages that says something like "Trying to decide how to deploy Sourcegraph? See our recommendations for how to deploy Sourcegraph for your needs" (obviously feel free to rework this wording).

That way we can add more information over time to the tables, add more commentary, etc, and the index page becomes the single source of truth on how to decide which deployment option to choose.

File: doc/admin/install/index.md [hunk 5]
```diff
@@ -1,4 +1,7 @@
...
+|-------------------------------------------------------|-----------------------------------------------------|------------|----------------|---------------|-------------|
+| [Single-container server](../install/docker/index.md) | Local testing                                       | 60 seconds | Impossible     | No            | No          |
+| [Docker Compose](../install/docker-compose/index.md)  | Small & medium production deployments               | 5 minutes  | Possible       | No            | Yes         |
+| [Kubernetes](../install/cluster.md)                   | Medium & large highly-available cluster deployments | 30 minutes | Easily         | Yes           | Yes         |
```
Comment (dadlerj) [inline doc/admin/install/index.md]: This line is confusing, as "Cluster" represents two different types of deployments, not just Kubernetes. Perhaps the title for this one should be "Cluster"?

Comment (ggilmore) [reply doc/admin/install/index.md]: Well, the pure docker doesn't actually support all the features in this table (auto-healing, service discovery etc.). Maybe they both need their own pages, but I think that's out of scope for this PR.

**Pull request merged**